lukasjarosch/skipper
# Feat/gcp kms support

Implementation of https://github.com/lukasjarosch/skipper/issues/46

Adds the GCP KMS Secrets Driver. To use it add following to your `<target>.skipper.secrets.drivers` target configuration:

```yaml
gcpkms:
  key_name: "projects/<project>/locations/<location>/keyRings/<keyring>/cryptoKeys/<keyname>"
```

The key name represents the full key URI. The naming scheme might be confusing but is consistent with the SDK. Both the URI and the key name are called name. I did not (yet) implement the option to provide project, location, key ring, etc. as dedicated arguments because as opposed to the Azure SDK, the GCP SDK wants to have the URI for the request. Let me know if you think the single arguments implementation should be included as well.

### Open issues

**Fixed**: ~~I'm still struggling with key versions. With Cloud KMS, rotating a key does not disable older versions and you can still decrypt secrets with the older version without actually needing to specify the version, even if you changed the primary (default) version. Somehow that just works until you disable the old version.~~

~~In theory you should be able to specify the key version by appending it to the URI: `projects/<project>/locations/<location>/keyRings/<keyring>/cryptoKeys/<keyname>/cryptoKeyVersions/<version>`. This is the URI for key versions, and supported within the SDK (as you can see [here](https://github.com/googleapis/google-cloud-go/blob/kms/v1.6.0/kms/apiv1/kmspb/service.pb.go#L1745)), but it the API throws me an error and I wasn't successful fixing this yet:~~

```
failed to replace secret value: failed to decrypt input: rpc error: code = InvalidArgument desc = projects/tmp-joe-skipper-test/locations/europe-west3/keyRings/test/cryptoKeys/quickstart/cryptoKeyVersions/2 is not a valid resource type for this request.
```

### Testing instructions

1. Create a new key ring (equivalent to Azure Key Vault, but only for keys) and a new key:

```bash
gcloud kms keyrings create "myring" --location "europe-west3"
gcloud kms keys create "mykey" --location "europe-west3" --keyring "myring" --purpose "encryption" 
```

2. Encrypt a secret:

```bash
printf "supersecret" | gcloud kms encrypt --key "mykey" --keyring "myring" --location "europe-west3" --plaintext-file=- --ciphertext-file=- | base64
```

3. Place that output as secret in your skipper inventory secret path:

```yaml
data: <encryption output>
type: gcpkms
```
4. Configure the GCP KMS Secret Driver in your target as stated above.

5. Make sure skipper [can authenticate itself against Google Cloud](https://cloud.google.com/docs/authentication/application-default-credentials) by either pointing to service account credentials with the`GOOGLE_APPLICATION_CREDENTIALS` environment variable or running `gcloud auth application-default login` to let skipper authenticate itself as you. As this is a test environment, just be the owner of the project to not worry about permissions.

6. Run skipper.
